contracts/risk/CoverageDataProviderV2.sol

```diff
@@ -0,0 +1,230 @@
+import "../interfaces/ISOLACE.sol";
+import "../utils/Governable.sol";
+
+/**
+ * @title  CoverageDataProviderV2
+ * @author solace.fi
+ * @notice Holds underwriting pool amounts in `USD`. Provides information to the [**Risk Manager**](./RiskManager.sol) that is the maximum amount of cover that `Solace` protocol can sell as a coverage.
+*/
+contract CoverageDataProviderV2 is ICoverageDataProviderV2, Governable, ReentrancyGuard {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /***************************************
+     STATE VARIABLES
+    ***************************************/
+
+    /// @notice The balance of underwriting pool in usd.
+    mapping(string => uint256) private _uwpBalanceOf;
+
+    /// @notice The index to underwriting pool.
+    mapping(uint256 => string) private _indexToUwp;
+
+    /// @notice The underwriting pool to index.
+    mapping(string => uint256) private _uwpToIndex;
+
+    /// @notice The underwriting pool updaters.
+    EnumerableSet.AddressSet private _updaters;
+
+    /// @notice The underwriting pool count
+    uint256 public numOfPools;
+
+    /***************************************
+     MODIFIERS FUNCTIONS
+    ***************************************/
+    
+    modifier canUpdate() {
+      require(msg.sender == super.governance() || isUpdater(msg.sender), "!governance");
+      _;
+    }
+
+    /**
+     * @notice Constructs the `CoverageDataProviderV2` contract.
+     * @param _governance The address of the [governor](/docs/protocol/governance).
+    */
+    // solhint-disable-next-line no-empty-blocks
+    constructor(address _governance) Governable(_governance) {}
+
+    /***************************************
+     MUTUATOR FUNCTIONS
+    ***************************************/
+   
+    /**
+      * @notice Resets the underwriting pool balances.
+      * @param _uwpNames The underwriting pool values to set.
+      * @param _amounts The underwriting pool balances in `USD`.
+    */
+    function set(string[] calldata _uwpNames, uint256[] calldata _amounts) external override nonReentrant canUpdate {
+      require(_uwpNames.length == _amounts.length, "length mismatch");
+      _set(_uwpNames, _amounts);
+    }
+
+    /**
+     * @notice Removes the given underwriting pool.
+     * @param uwpNames The underwriting pool names to remove.
+    */
+    function remove(string[] calldata uwpNames) external override canUpdate {
+      _remove(uwpNames);
+    }
+
+    /***************************************
+     VIEW FUNCTIONS
+    ***************************************/
+
+    /**
+     * @notice Returns the maximum amount of cover in `USD` that Solace as a whole can sell.
+     * @return cover The max amount of cover in `USD`.
+    */
+    function maxCover() external view override returns (uint256 cover) {
+      // get pool balance
+      uint256 pools = numOfPools;
+      for (uint256 i = pools; i > 0; i--) {
+        cover += balanceOf(_indexToUwp[i]);
+      }
+    }
```

Comment:
should this be calculated when balances are set?

Reply:
Yes good point. I refactored it.